### ISSUE TITLE
Utilize Modularity for Performance, Scalability and Reliability

### ISSUE THREAD

**Issue by dlapcevic** [cilium/cilium#33360]
## Problem

cilium-agent is excessively using node and K8s Control Plane resources (CPU, Memory, Network Throughput). It hinders its and the system's performance and scalability.

#### Why does the problem occur?
cilium-agent by default runs a lot of processes for the features that aren’t enabled or used.
K8s watchers are initialized despite not being used, or the features that require them may not be used at a time.

## Proposal
Utilize modularization (the hive system) to ensure that only required processes are running.

### Modularization

Every feature is a self contained cell, decoupled from other feature cells as much as possible.
Every feature is implemented as a cell that can be started and stopped independently of other feature cells.

### Feature Enablement Framework

**Enablement**
A feature cell is started when a feature is enabled and stopped when a feature is disabled.
A feature cell can be restarted any number of times without side effects.

**K8s Resource Tracking**
Track which feature cells are running and the required K8s API resources for those cells.
Watchers for K8s API resources that aren’t required by any features will be disabled.

### Dynamic Feature Enablement

Allow cilium-agent and cilium-operator to read configuration changes while running, and not to require restart.

**High Level Design**
A new ConfigMap cilium-dynamic-config will be created.
cilium-agent watches the new ConfigMap and dynamically starts and stops cells.
cilium-operator watches the base ConfigMap cilium-config and other K8s resources, determines which features are enabled and writes it to the cilium-dynamic-config.

#### Alternative

Use mounted volumes. Read from files. Delay of changes to ConfigMap will be 1 minute. Can it be reduced?
No advantages, since kubelet anyway creates a watcher for each mounted ConfigMap.

### Lazy Feature Initialization

A feature cell is initialized only after it’s used, despite the feature being enabled.
(e.g. network policy enforcement system is only initialized after a network policy is created in the cluster)

## End goal example

1. Network policy feature is disabled based on a flag in cilium-config.
2. Network policy cell isn’t started. Cilium Endpoint, Cilium Identity and Cilium Endpoint Slice aren’t managed (created, updated or 3. deleted).
4. The watchers for the same K8s API resources aren’t started.
5. Network policy feature is enabled when the flag in cilium-config is switched on.
6. cilium-operator reads the changed cilium-config, but doesn’t make any actions because there are no network policies in the cluster.
7. A network policy resource is created in the cluster.
8. cilium-operator updates the dynamic ConfigMap.
9. cilium-agent reads it and starts the network policy cell.
10. cilium-agent starts the required watchers.
11. Network policy enforcement system is enabled.

## Dependencies

Dependency for network policy modularization: Modularization: pkg/identity https://github.com/cilium/cilium/issues/23767

## List of Features

Additional features that are suitable to use the described optimizations will be added to the list over time.

**Comment by dlapcevic** [comment]
cc @ovidiutirla @sypakine @joestringer @marseel

**Comment by marseel** [comment]
Some quick thoughts:

Modularization:
on high-level of enabling/disabling cells definitely makes sense. I think we already do it, at least for new features, for example l2announcement: https://github.com/cilium/cilium/blob/4219e5c3292612b7ecd221d39a683de1996ce71a/pkg/l2announcer/l2announcer.go#L122
so it's the matter of modularizing "old" features.

There might be gaps when a feature is disabled, like for example we create CRD in the operator - but that's something to improve on. I don't think we start watchers for disabled features / unused CRDs: https://github.com/cilium/cilium/blob/main/pkg/k8s/resource/resource.go#L462

Dependant features:
Theoretically we have graph dependency in the hive, so for example, if someone wants to enable EGW policy, we could determine that CEP and CID is also required and turn that on too, but it sounds a bit tricky to automatically ensure that all dependant features are on too. 

For dynamic enablement:
I'm not sure about it. The usual way of enabling a feature is through rolling upgrade, which is definitely much safer operation than turning the feature at the same time in all agents - which would essentially break things cluster-level while in rolling upgrade it would potentially stop on the first agent failing. 

There might be cases in datapath as well, where enabling particular feature needs recreation of ebp programs/maps and I think that currently we perform it safely only on startup - but I am far from the expert in this area, so would appreciate someone else input. 

So for dynamic enablement sounds like we would need some orchestrator that enables it per agent, which would essentially be similar to the rolling upgrade done by k8s. 

For NetworkPolicy example:
Synchronized starting of all watchers sounds like a significant spike load. Also, assuming that CEPs are not created, creating them with the first NetworkPolicy would be significant load too on control-plane.

Alternatively, enabling NetworkPolicies through regular config-map and performing rolling upgrade sounds like more predictable load on control plane.

**Comment by joestringer** [comment]
Modularization makes sense, already tracked by this issue: https://github.com/cilium/cilium/issues/23425 . 

Dynamic enablement is conceptually interesting. As Marcel points out, it would be helpful to have a model of understanding about how invasive the change is and how we appropriately stage those changes to minimize impact. Not all changes can be safely transitioned, for instance the spike load case. Also worth mentioning that sometimes certain changes are implemented in Helm at the moment with changes to other Cilium resources, not just the ConfigMap. Depending on the functionality, maybe we need to modify other settings, change capabilities, or otherwise configure multiple resources. Dynamically handling ConfigMap changes is just one piece of the puzzle. Related, I think it would help to have a few really good use cases to demonstrate the current pain points and how this proposal would improve those situations from a user perspective.

The network policy example doesn't make sense to me at all. I feel like we'd need to take a step back to consider the high level architecture of Cilium, the functionality it provides, and the personas who we are trying to design this system for.

**Comment by dlapcevic** [comment]
Thanks for the feedback.

I understand that network policy is one of the main and most useful features of Cilium, and I agree that ensuring network security is very valuable. Therefore, perhaps network policies are not a good example that I included in the description. However, I can definitely see how Cilium can be used for a wide variety of its amazing features, without enabling all / most of the features. Designing the system in a way that will allow supporting maximum scalability and minimum footprint for any specific configuration brings great benefit.

**Comment by dlapcevic** [comment]
Hi, following up on this.

I wanted to provide more information behind the motivation for trying to move in this direction.

We want to modularize network policy enforcement system because we have users on GKE that require very high scale (# of nodes and pod churn rate) which currently isn’t within the scalability limits of Cilium network policies. We want network policies to be enabled eventually, by making them more scalable with the other proposed improvements. Right now, we have a choice to either disable network policies in Cilium, or to disable Cilium entirely if we cannot only disable network policies. I would prefer to disable only network policies temporarily until the required scalability is achieved.

Besides strictly requiring disabling features, the way I see we can achieve this for all users, in the most user friendly way, is to modularize and lazily initialize parts of the system, so that Cilium can scale up to it’s maximum for any feature set.

**Comment by dlapcevic** [comment]
Thank you @learnitall for the detailed explanation.

I agree with you. Your insights help me better understand the direction we want to take, one that fits in and benefits different aspects both from developer and user sides.

Perhaps before moving forward with the Lazy Initialization and related discussions, I want to see how I can contribute to modularization. I've assigned the Modularization: pkg/identity issue to myself and I'll start from there. https://github.com/cilium/cilium/issues/23767

I’ve taken a look into Audit Policy (flag `policy-audit-mode`) and as far as I see it doesn’t disable the network policy enforcement system, specifically Cilium Endpoints and Cilium Identities, which presents the scalability bottleneck.
We’ve been able to run Cilium on GKE on clusters with up to 5k nodes, 200k pods, 100/s pod churn rate, and there is a similar ongoing effort here https://github.com/cilium/cilium/issues/33778.
Another thing we want to do is to provide data about the issues that Cilium faces at such and higher scales, so that we can together start tackling these scalability obstacles, and plan for having network policies on even higher scales. The two most relevant scalability dimensions are number of nodes in the cluster and pod churn rate, and there are users that want to stretch both of these dimensions.

**Comment by hemanthmalla** [comment]
@dlapcevic In case you haven't seen already, does the feature set provided by [CiliumNodeConfig](https://docs.cilium.io/en/stable/configuration/per-node-config/#per-node-configuration) align with what you're looking for with `cilium-dynamic-config` ?

**Comment by dlapcevic** [comment]
> [@dlapcevic](https://github.com/dlapcevic) In case you haven't seen already, does the feature set provided by [CiliumNodeConfig](https://docs.cilium.io/en/stable/configuration/per-node-config/#per-node-configuration) align with what you're looking for with `cilium-dynamic-config` ?

Hi @hemanthmalla, thank you for the suggestion. I think this is useful, but we still need to restart cilium-agents to pick up the config. The initial proposal for `cilium-dynamic-config` was to avoid cilium-agent restart, by picking up config while running and starting/stopping parts of the system according to the overridden config, likely using hive/cell.

Anyway, I think we could use per node configuration to move a step closer to dynamically enabling features, but still with cilium-agent restart. We'll need a mechanism that will control rollouts of features. This would be a solution until we have a more modular system where we can clearly decouple features and easily enable them. We're not so close to it yet, and we don't have a clear ETA for it.
